File: static/modules/version_info.js
import { apiJson, esc } from "./api.js";

let _bound = false;
let _appMeta = null;

function formatReleaseDate(value) {
  const text = String(value || "").trim();
  if (!text) return "";
  return text.replaceAll("-", "/");
}

function buildSectionHtml(section) {
  const title = esc(section?.title || "");
  const items = Array.isArray(section?.items) ? section.items : [];
  const itemHtml = items.map(item => `<li>${esc(item || "")}</li>`).join("");
  return `
    <section class="version-section">
      <h4>${title}</h4>
      <ul>${itemHtml}</ul>
    </section>`;
}

function renderVersionInfo(meta) {
  const versionButton = document.getElementById("btn-app-version");
  const versionLabel = document.getElementById("app-version-label");
  const modal = document.getElementById("version-info-modal");
  const title = document.getElementById("version-info-title");
  const subhead = document.getElementById("version-info-subhead");
  const content = document.getElementById("version-info-content");

  if (!versionButton || !versionLabel || !modal || !title || !subhead || !content) return;

  const version = String(meta?.version || "").trim();
  const appName = String(meta?.app_name || "").trim() || "OpenText 出貨排程系統";
  const releaseDate = formatReleaseDate(meta?.released_at);
  const headline = String(meta?.headline || "").trim();
  const sections = Array.isArray(meta?.sections) ? meta.sections : [];

  versionLabel.textContent = version || "版本資訊";
  versionButton.style.display = "inline-flex";
  versionButton.title = version ? `${version} 更新內容` : "更新內容";

  title.textContent = version ? `${appName} ${version}` : appName;
  subhead.textContent = [releaseDate, headline].filter(Boolean).join("  |  ");
  content.innerHTML = sections.map(buildSectionHtml).join("");

  if (version) {
    document.title = `${appName} ${version}`;
  } else {
    document.title = appName;
  }
}

function openVersionInfoModal() {
  const modal = document.getElementById("version-info-modal");
  if (!modal || !_appMeta) return;
  modal.style.display = "flex";
}

function closeVersionInfoModal() {
  const modal = document.getElementById("version-info-modal");
  if (!modal) return;
  modal.style.display = "none";
}

function bindVersionInfoEvents() {
  if (_bound) return;
  _bound = true;

  document.getElementById("btn-app-version")?.addEventListener("click", openVersionInfoModal);
  document.getElementById("version-info-close")?.addEventListener("click", closeVersionInfoModal);
  document.getElementById("version-info-ok")?.addEventListener("click", closeVersionInfoModal);
  document.getElementById("version-info-modal")?.addEventListener("click", event => {
    if (event.target?.id === "version-info-modal") {
      closeVersionInfoModal();
    }
  });
}

export async function initVersionInfo() {
  bindVersionInfoEvents();
  _appMeta = await apiJson("/api/system/app-meta");
  renderVersionInfo(_appMeta);
  return _appMeta;
}

File: app/version_info.py
from __future__ import annotations

APP_NAME = "出貨排程系統"
APP_VERSION = "v2026.04.08.2"
APP_RELEASED_AT = "2026-04-08"
APP_HEADLINE = "這版修正缺料 modal 的舊補料還原邏輯，後面訂單不會再把自己以前存過的補料值硬帶回來。"
APP_CHANGELOG = [
    {
        "title": "補料還原",
        "items": [
            "缺料 modal 重新開啟時，現在會用「訂單 + 料號」還原先前的補料，不再把同料號不同訂單的舊值混在一起。",
            "像 2-5 補了 PB-20135A-TAB 之後，2-7 不會再因為自己以前存過的補料值被硬撐出來。",
            "當前面訂單已補平後，後面同料號卡片若已沒有缺口，會直接隱藏，不再因為舊補料值而留在畫面上。",
        ],
    },
    {
        "title": "建議補料",
        "items": [
            "同料號若後面訂單還會繼續吃料，前一張卡的「建議補」會直接帶入後續累積缺口，不再只看自己這一筆。",
            "像 2-5 的 EC-20131A-TAB，現在會直接建議補 2000，不會只先開 1000 又讓 2-6 再跳一次。",
            "右側缺料區和缺料 modal 都會套用同一套往後看的建議量算法，畫面判讀會更一致。",
        ],
    },
    {
        "title": "順序扣帳",
        "items": [
            "缺料 modal 仍然保留逐筆補料，但同料號會照訂單順序往後重算 running balance。",
            "前面一筆已經補到足夠時，後面同料號會自動被補平並隱藏，不會再讓人誤以為要重複補。",
        ],
    },
]


def get_app_meta() -> dict:
    return {
        "app_name": APP_NAME,
        "version": APP_VERSION,
        "released_at": APP_RELEASED_AT,
        "headline": APP_HEADLINE,
        "sections": APP_CHANGELOG,
    }
